external-crates/move/crates/move-compiler/src/hlir/match_compilation.rs

```diff
@@ -189,8 +218,66 @@ impl PatternArm {
                 }
                 Some((vec![], output))
             }
-            TP::Constructor(_, _, _, _, _) | TP::BorrowConstructor(_, _, _, _, _, _) => None,
+            TP::Variant(_, _, _, _, _) | TP::BorrowVariant(_, _, _, _, _, _) => None,
+            TP::Struct(_, _, _, _) | TP::BorrowStruct(_, _, _, _, _) => None,
+            TP::Literal(_) => None,
+            TP::Binder(mut_, x) => {
+                for arg_type in arg_types
+                    .clone()
+                    .into_iter()
+                    .map(|ty| ty_to_wildcard_pattern(ty.clone(), loc))
+                    .rev()
+                {
+                    output.pats.push_front(arg_type);
+                }
+                Some((vec![(mut_, x)], output))
+            }
+            TP::Wildcard => {
+                for arg_type in arg_types
+                    .clone()
+                    .into_iter()
+                    .map(|ty| ty_to_wildcard_pattern(ty.clone(), loc))
+                    .rev()
+                {
+                    output.pats.push_front(arg_type);
+                }
+                Some((vec![], output))
+            }
+            TP::Or(_, _) => unreachable!(),
+            TP::At(x, inner) => {
+                output.pats.push_front(*inner);
+                output
+                    .specialize_variant(context, ctor_name, arg_types)
+                    .map(|(mut binders, inner)| {
+                        binders.push((Mutability::Imm, x));
+                        (binders, inner)
+                    })
+            }
+            TP::ErrorPat => None,
+        }
+    }
+
+    fn specialize_struct(
```

Comment:
Combining `speicalize_struct` / `specialize_variant` is an interesting idea. I see why you would suggest it: the logic largely overlaps. I am worried about doing it from a sanity perspective, though: if for some reason something went wrong earlier in the compiler and you ended up with both a variant and a struct in the same column, you would get incredibly strange crashes.

In addition, since the rest of the code needs to treat them separately anyway, this seems like a nice parity. (It also makes debugging specific cases easier...)